datumaro/components/operations.py

```diff
@@ -106,40 +108,95 @@ def merge(cls, *sources):
     def merge_items(cls, existing_item, current_item):
         return existing_item.wrap(
             image=cls.merge_images(existing_item, current_item),
+            attributes=cls.merge_attrs(
+                existing_item.attributes, current_item.attributes,
+                item_id=(existing_item.id, existing_item.subset)),
             annotations=cls.merge_anno(
                 existing_item.annotations, current_item.annotations))
 
     @staticmethod
-    def merge_images(existing_item, current_item):
-        image = None
-        if existing_item.has_image and current_item.has_image:
-            if existing_item.image.has_data:
-                image = existing_item.image
+    def merge_attrs(a: Dict, b: Dict,
+            item_id: Optional[Tuple[str, str]] = None) -> Dict:
+        merged = {}
+
+        for name in set(a) | set(b):
```

Comment:
```suggestion
        for name in a.keys() | b.keys():
```

This avoids creation of new sets.

Reply:
I didn't think dict views are documented to be set-like, but it seems, they are.

Reply:
It is documented here: https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects.

Reply:
Done

Reply:
I think you forgot to push it.